HiddenFieldCheck: parameter shadowing top-level field not flagged in compact source files (JEP 512)

Parent Issue #19971

I have read check documentation: https://checkstyle.org/checks/coding/hiddenfield.html
I have downloaded the latest checkstyle from: https://checkstyle.org/cmdline.html#Download_and_Run
I have executed the cli and showed it below, as cli describes the problem better than 1,000 words

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > config.xml <<'EOF'
<?xml version="1.0"?>
<!DOCTYPE module PUBLIC
    "-//Checkstyle//DTD Checkstyle Configuration 1.3//EN"
    "https://checkstyle.org/dtds/configuration_1_3.dtd">
<module name="Checker">
  <module name="TreeWalker">
    <module name="HiddenField"/>
  </module>
</module>
EOF

vivek@Viveks-MacBook-Air checkstyle % cat > InputHiddenFieldCompact.java <<'EOF'
String name = "default";
int count = 0;
void process(String name, int count) {
    System.out.println(name + ": " + count);
}
void main() { process("test", 42); }
EOF

vivek@Viveks-MacBook-Air checkstyle % RUN_LOCALE="-Duser.language=en -Duser.country=US"
java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml InputHiddenFieldCompact.java

Starting audit...
Audit done.
```

Same code inside an ordinary class:

```bash
vivek@Viveks-MacBook-Air checkstyle % cat > OrdinaryHiddenField.java <<'EOF'
public class OrdinaryHiddenField {
    String name = "default";
    int count = 0;
    void process(String name, int count) {
        System.out.println(name + ": " + count);
    }
}
EOF

vivek@Viveks-MacBook-Air checkstyle % java $RUN_LOCALE -jar target/checkstyle-13.5.0-SNAPSHOT-all.jar -c config.xml OrdinaryHiddenField.java

Starting audit...
[ERROR] OrdinaryHiddenField.java:4:25: 'name' hides a field. [HiddenField]
[ERROR] OrdinaryHiddenField.java:4:35: 'count' hides a field. [HiddenField]
Audit done.
Checkstyle ends with 2 errors.
```

---

**Describe what you expect in detail.**

HiddenFieldCheck must detect a method parameter that shadows a top-level field in a JEP 512 compact source file. The check builds its `FieldFrame` by visiting `CLASS_DEF`, `ENUM_DEF`, and `RECORD_DEF` tokens and collecting field declarations from their bodies. A compact source file's implicit class is represented by none of these tokens, so top-level fields are never registered in any `FieldFrame`, and parameter shadowing goes undetected.

Expected on the compact input: 2 violations (`'name' hides a field.` and `'count' hides a field.`), identical to the ordinary class above.

--- a/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheck.java
+++ b/src/main/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheck.java
@@ -25,6 +25,8 @@ import java.util.Objects;
 import java.util.Set;
 import java.util.regex.Pattern;
 
+import javax.annotation.Nullable;
+
 import com.puppycrawl.tools.checkstyle.FileStatefulCheck;
 import com.puppycrawl.tools.checkstyle.api.AbstractCheck;
 import com.puppycrawl.tools.checkstyle.api.DetailAST;
@@ -135,6 +137,7 @@ public class HiddenFieldCheck
             TokenTypes.LAMBDA,
             TokenTypes.RECORD_DEF,
             TokenTypes.RECORD_COMPONENT_DEF,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
         };
     }
 
@@ -145,6 +148,7 @@ public class HiddenFieldCheck
             TokenTypes.ENUM_DEF,
             TokenTypes.ENUM_CONSTANT_DEF,
             TokenTypes.RECORD_DEF,
+            TokenTypes.COMPACT_COMPILATION_UNIT,
         };
     }
 
@@ -216,7 +220,7 @@ public class HiddenFieldCheck
         final FieldFrame newFrame = new FieldFrame(frame, isStaticInnerType, frameName);
 
         // add fields to container
-        final DetailAST objBlock = ast.findFirstToken(TokenTypes.OBJBLOCK);
+        final DetailAST objBlock = getFieldContainer(ast);
         // enum constants may not have bodies
         if (objBlock != null) {
             DetailAST child = objBlock.getFirstChild();
@@ -249,6 +253,25 @@ public class HiddenFieldCheck
         }
         // push container
         frame = newFrame;
+    }
+
+    /**
+     * Gets the member container for field declaration harvesting.
+     *
+     * @param ast the type definition node.
+     * @return the member container, either the compact compilation unit
+     *     itself or the OBJBLOCK child of a standard type definition.
+     */
+    @Nullable
+    private static DetailAST getFieldContainer(DetailAST ast) {
+        final DetailAST result;
+        if (ast.getType() == TokenTypes.COMPACT_COMPILATION_UNIT) {
+            result = ast;
+        }
+        else {
+            result = ast.findFirstToken(TokenTypes.OBJBLOCK);
+        }
+        return result;
     }
 
     @Override

--- a/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheckTest.java
+++ b/src/test/java/com/puppycrawl/tools/checkstyle/checks/coding/HiddenFieldCheckTest.java
@@ -684,6 +684,53 @@ public class HiddenFieldCheckTest
     }
 
     @Test
+    public void testHiddenFieldCompactSourceFile() throws Exception {
+        final String[] expected = {
+            "18:21: " + getCheckMessage(MSG_KEY, "name"),
+            "18:31: " + getCheckMessage(MSG_KEY, "count"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputHiddenFieldCompactSourceFile.java"), expected);
+    }
+
+    @Test
+    public void testHiddenFieldCompactSourceFileStatic() throws Exception {
+        final String[] expected = {
+            "17:21: " + getCheckMessage(MSG_KEY, "name"),
+        };
+        verifyWithInlineConfigParser(
+                getNonCompilablePath("InputHiddenFieldCompactSourceFileStatic.java"), expected);
+    }
+
+    @Test
+    public void testHiddenFieldCompactSourceFileStaticFieldAndMethod() throws Exception {
+        final String[] expected = {
+            "17:28: " + getCheckMessage(MSG_KEY, "name"),
+        };
+        final String file = "InputHiddenFieldCompactSourceFileStaticFieldAndMethod.java";
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(file), expected);
+    }
+
+    @Test
+    public void testHiddenFieldCompactSourceFileInstanceFieldStaticMethod() throws Exception {
+        final String[] expected = {};
+        final String file = "InputHiddenFieldCompactSourceFileInstanceFieldStaticMethod.java";
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(file), expected);
+    }
+
+    @Test
+    public void testHiddenFieldCompactSourceFileNested() throws Exception {
+        final String[] expected = {
+            "24:25: " + getCheckMessage(MSG_KEY, "name"),
+        };
+        final String file = "InputHiddenFieldCompactSourceFileNested.java";
+        verifyWithInlineConfigParser(
+                getNonCompilablePath(file), expected);
+    }
+
+    @Test
     public void testHiddenFieldRecordPattern() throws Exception {
 
         final String[] expected = {

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFile.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFile.java
@@ -1,0 +1,25 @@
+/*
+HiddenField
+ignoreFormat = (default)null
+ignoreConstructorParameter = (default)false
+ignoreSetter = (default)false
+setterCanReturnItsClass = (default)false
+ignoreAbstractMethods = (default)false
+tokens = VARIABLE_DEF, PARAMETER_DEF
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+String name = "default";
+int count = 0;
+
+void process(String name, int count) {
+    // 2 violations above:
+    //                   ''name' hides a field.'
+    //                   ''count' hides a field.'
+    System.out.println(name + ": " + count);
+}
+
+void main() { process("test", 42); }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileInstanceFieldStaticMethod.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileInstanceFieldStaticMethod.java
@@ -1,0 +1,21 @@
+/*
+HiddenField
+ignoreFormat = (default)null
+ignoreConstructorParameter = (default)false
+ignoreSetter = (default)false
+setterCanReturnItsClass = (default)false
+ignoreAbstractMethods = (default)false
+tokens = VARIABLE_DEF, PARAMETER_DEF
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+String name = "default";
+
+static void process(String name) {
+    System.out.println(name);
+}
+
+void main() { process("test"); }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileNested.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileNested.java
@@ -1,0 +1,25 @@
+/*
+HiddenField
+ignoreFormat = (default)null
+ignoreConstructorParameter = (default)false
+ignoreSetter = (default)false
+setterCanReturnItsClass = (default)false
+ignoreAbstractMethods = (default)false
+tokens = VARIABLE_DEF, PARAMETER_DEF
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+String name = "default";
+
+void main() { }
+
+static class StaticNested {
+    void process(String name) { }
+}
+
+class Inner {
+    void process(String name) { } // violation ''name' hides a field'
+}

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileStatic.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileStatic.java
@@ -1,0 +1,21 @@
+/*
+HiddenField
+ignoreFormat = (default)null
+ignoreConstructorParameter = (default)false
+ignoreSetter = (default)false
+setterCanReturnItsClass = (default)false
+ignoreAbstractMethods = (default)false
+tokens = VARIABLE_DEF, PARAMETER_DEF
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+static String name = "default";
+
+void process(String name) { // violation, ''name' hides a field'
+    System.out.println(name);
+}
+
+void main() { process("test"); }

--- a/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileStaticFieldAndMethod.java
+++ b/src/test/resources-noncompilable/com/puppycrawl/tools/checkstyle/checks/coding/hiddenfield/InputHiddenFieldCompactSourceFileStaticFieldAndMethod.java
@@ -1,0 +1,21 @@
+/*
+HiddenField
+ignoreFormat = (default)null
+ignoreConstructorParameter = (default)false
+ignoreSetter = (default)false
+setterCanReturnItsClass = (default)false
+ignoreAbstractMethods = (default)false
+tokens = VARIABLE_DEF, PARAMETER_DEF
+
+
+*/
+
+// non-compiled with javac: Compilable with Java25
+
+static String name = "default";
+
+static void process(String name) { // violation ''name' hides a field'
+    System.out.println(name);
+}
+
+void main() { process("test"); }

